### PR TITLE
WIP: Refactor LRSDP to support more general linear constraints

### PR DESCRIPTION
**Pull request by epatters**
This PR is intended to address #578.

It is *not* ready to merge. I'm starting the PR to get some preliminary feedback and track my progress.

### CODE REVIEW

**Comment by epatters** [comment]
I've sketched an interface and implementations for a `LinearConstraint` class. It needs to be tested and integrated with `SDP` and `LRSDP`.

Any comments are welcome.

File: src/mlpack/core/optimizers/sdp/linear_constraint_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+namespace mlpack {
+namespace optimization {
+
+// MatrixConstraint implementation
+
+template <typename MatrixType>
+arma::vec MatrixConstraint<MatrixType>::Evaluate(const arma::mat& x)
+{
+  const size_t m = _a.size();
+  arma::vec vals = arma::vec(m);
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sdp/linear_constraint_impl.hpp]: You can just do `arma::vec vals(m);` or even `arma::vec vals(_a.size());` here.

File: src/mlpack/core/optimizers/sdp/linear_constraint.hpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  //! Return the vector of b values.
+  const arma::vec& B() const { return _b; }
+  
+  //! Modify the vector of b values.
+  arma::vec& B() { return _b; }
+
+private:
+  std::vector<MatrixType> _a;
+  arma::vec _b;
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sdp/linear_constraint.hpp]: Could you change these member names to just `a` and `b` please?  See https://github.com/mlpack/mlpack/wiki/DesignGuidelines#naming-conventions (this comment applies in other places too, I think).

**Comment by rcurtin** [comment]
Hey, thanks for the PR.  This looks like a good start.  I'm interested in how the SDP class will work with this since there are three methods `Evaluate()`, `EvaluateLR()` and `GradientLR()`.  I guess I had thought that each constraint would provide just an `Evaluate()` method which would take the current coordinate matrix X as a parameter, and then return the constraint value.  So `SDP::EvaluateConstraint(i)` would just forward to constraint i's `Evaluate()` method.  But that doesn't seem to be the approach here (just guessing)---maybe I have overlooked something?

**Comment by epatters** [comment]
@rcurtin: Thanks for the comments. I'll make these changes. Also, I didn't know about the design guidelines. That's a helpful resource.

The reason for the complicated API is efficiency. Constraints of the same type are grouped together because it sometimes allow efficiency gains. For example, look at the current version of `GradientImpl` in LRSDP. In computing `L = 2*S*R`, it first computes `S` (involving all the constraints), then multiplies by `R`. If each individual constraint had its own gradient function, they would each have to multiply by `R`, leading to many unnecessary matrix multiplications. On the other hand, a different computation strategy is appropriate for low-rank matrices.

As for the evaluate functions, specializing to `EvaluateLR` is important because the matrix RR^T should really never be formed at all. (There are some TODOs in the current implementation to that effect.)  That said, at present there is no advantage to the grouping the evaluations, so I suppose I could separate out those for the individual constraints.

**Comment by rcurtin** [comment]
Hey Evan,

Sorry for the slow response.  Things have been busy and I haven't had a chance to look further into this until now.

I wonder, should we even be including an `Evaluate()` and a `Gradient()` function in these constraint types?  Or instead should we simply be allowing access to the A matrix (however it should be represented) and the b vector, and allow the optimizers to do what they need with those?  (This is closer to how the current code works, but the current code has a lot of template specializations that I'd love to do away with, favoring just one non-specialized implementation instead.)

This is simple for the sparse and dense constraints, which can just return A as types `arma::sp_mat` and `arma::mat`, respectively, but the linear constraint is a bit harder because we want to represent A as (aa').  We could possibly use the Armadillo type that represents two vectors to be multiplied for this... it'd look something like `arma::glue<arma::Col<eT>, arma::Col<eT>, arma::op_multiply>`.  If we return that type for the A matrix and then use that in a larger Armadillo expression, I think we can reasonably expect that the template metaprogramming that evaluates Armadillo expressions will do a decent job.

Is what I'm saying making any sense?  Please let me know if not.  It has been a while since I have thought about SDPs.  My concern though is that ideally we don't want to include to much optimizer-specific logic in the constraints, and it seems like in this case that's what's happening with `EvaluateLR()` and `GradientLR()`; if we were to come up with another technique for solving SDPs, like for instance with block diagonal representations instead of low-rank, we would probably need an `EvaluateBlockDiagonal()` and `GradientBlockDiagonal()` in each constraint, which I'd like to avoid.

**Comment by rcurtin** [comment]
Hi @epatters, I'm going to go ahead and close this now for inactivity.  If you ever have the time or interest to revisit this, I am happy to help where I can.

**Pull request closed without merging**